### PR TITLE
feat: add plugin via api

### LINKED ISSUE #4236: Feature: Adding plugin via API support
### 🔖 Feature description

There should be a feature that should enable us to add plugin via API instead of sql scripts

### 🎤 Pitch

Adding sql script, for everytime we are adding a plugin is a cumbersome task, and there are so many relations that one needs to take care of that causes confusion sometimes while adding them, hence there should be an API support for adding plugins.

### ✌️ Solution

Write an API that would take all necessary details, for plugin creation and creates a plugin for me.

### 🔄️ Alternative

_No response_

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

AB#6767

### PR DESCRIPTION
**Pull request by prakash100198**
# Description

This pr aims at new feature where a user can add new plugin via api.
Fixes #4236 

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [x] I have commented my code, particularly in hard-to-understand areas.
* [x] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: api/restHandler/GlobalPluginRestHandler.go [hunk 1]
```diff
@@ -37,6 +45,96 @@ type GlobalPluginRestHandlerImpl struct {
...
+	var pluginDataDto plugin.PluginMetadataDto
+	err = decoder.Decode(&pluginDataDto)
+	if err != nil {
+		handler.logger.Errorw("request err, PatchPlugin", "error", err, "payload", pluginDataDto)
+		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+		return
+	}
+	handler.logger.Infow("request payload received for patching plugins")
+	// RBAC enforcer applying
+	isSuperAdmin, err := handler.userService.IsSuperAdmin(int(userId))
```
Comment (vikramdevtron) [inline api/restHandler/GlobalPluginRestHandler.go:67]: we can use casbin for super admin rbac

File: client/gitSensor/GitSensorClient.go [hunk 2]
```diff
@@ -63,8 +63,8 @@ func NewGitSensorClient(logger *zap.SugaredLogger, config *ClientConfig) (*Clien
 }
 
 type ClientConfig struct {
```
Comment (vikramdevtron) [inline client/gitSensor/GitSensorClient.go:65]: are these changes required

File: pkg/plugin/bean.go [hunk 3]
```diff
@@ -15,25 +17,44 @@ type PluginListComponentDto struct { //created new struct for backward compatibi
...
+	Description string            `json:"description"`
+	Type        string            `json:"type"` // SHARED, PRESET etc
+	Icon        string            `json:"icon"`
+	Tags        []string          `json:"tags"`
+	Action      int               `json:"action"`
+	PluginStage string            `json:"pluginStage,omitempty"`
+	PluginSteps []*PluginStepsDto `json:"pluginSteps,omitempty"`
+}
+
+type PluginStepsDto struct {
```
Comment (vikramdevtron) [inline pkg/plugin/bean.go:31]: do we have similar DTO at other places, we can use that

File: api/restHandler/GlobalPluginRestHandler.go (hunk 1 above)
Comment (prakash100198) [reply api/restHandler/GlobalPluginRestHandler.go:67]: internally it uses casbin to check for super admin

File: pkg/plugin/GlobalPluginService.go [hunk 4]
```diff
@@ -294,3 +301,1107 @@ func getVariableDto(pluginVariable *repository.PluginStepVariable) *PluginVariab
...
+
+func (impl *GlobalPluginServiceImpl) deletePlugin(pluginDeleteReq *PluginMetadataDto, userId int32) (*PluginMetadataDto, error) {
+	dbConnection := impl.globalPluginRepository.GetConnection()
+	tx, err := dbConnection.Begin()
+	if err != nil {
+		return nil, err
+	}
+	// Rollback tx on error
+	defer tx.Rollback()
+	pluginMetaData, err := impl.globalPluginRepository.GetMetaDataByPluginId(pluginDeleteReq.Id)
```
Comment (kartik-579) [inline pkg/plugin/GlobalPluginService.go:1335]: instead of directly starting deletion, let's check that if this plugin is referenced in some step or not. If it's being used we should throw error else ci/cd operations might break

**Review by kartik-579 — CHANGES_REQUESTED** [review]

File: pkg/plugin/GlobalPluginService.go [hunk 5]
```diff
@@ -294,3 +306,1137 @@ func getVariableDto(pluginVariable *repository.PluginStepVariable) *PluginVariab
...
+			newPluginTag := &repository.PluginTag{
+				Name: pluginTagReq,
+				AuditLog: sql.AuditLog{
+					CreatedOn: time.Now(),
+					CreatedBy: userId,
+					UpdatedOn: time.Now(),
+					UpdatedBy: userId,
+				},
+			}
+			newPluginTag, err = impl.globalPluginRepository.SavePluginTag(newPluginTag, tx)
```
Comment (kartik-579) [inline pkg/plugin/GlobalPluginService.go:438]: can do this in bulk and move mapping out of this iteration

File: pkg/plugin/GlobalPluginService.go [hunk 6]
```diff
@@ -294,3 +306,1137 @@ func getVariableDto(pluginVariable *repository.PluginStepVariable) *PluginVariab
...
+	if err != nil {
+		impl.logger.Errorw("error in getting all plugin tags", "err", err)
+		return err
+	}
+	//check for new tags, then create new plugin_tag and plugin_tag_relation entry in db when new tags are present in request
+	newPluginTagRelationsToCreate := make([]*repository.PluginTagRelation, 0)
+
+	for _, pluginTagReq := range pluginReq.Tags {
+		tagAlreadyExists := false
+		for _, presentPluginTags := range allPluginTags {
```
Comment (kartik-579) [inline pkg/plugin/GlobalPluginService.go:423]: can make this a map

**Review by kartik-579 — CHANGES_REQUESTED** [review]

**Review by kartik-579 — APPROVED** [review]

File: api/restHandler/GlobalPluginRestHandler.go [hunk 7]
```diff
@@ -37,6 +44,97 @@ type GlobalPluginRestHandlerImpl struct {
...
+func (handler *GlobalPluginRestHandlerImpl) GetDetailedPluginInfoByPluginId(w http.ResponseWriter, r *http.Request) {
+	userId, err := handler.userService.GetLoggedInUser(r)
+	if userId == 0 || err != nil {
+		common.WriteJsonResp(w, err, "Unauthorized User", http.StatusUnauthorized)
+		return
+	}
+	vars := mux.Vars(r)
+	pluginId, err := strconv.Atoi(vars["pluginId"])
+	if err != nil {
+		handler.logger.Errorw("error in converting from string to integer", "userId", userId, "err", err)
```
Comment (kripanshdevtron) [inline api/restHandler/GlobalPluginRestHandler.go:93]: log `pluginId` string value, would help in debugging

File: pkg/plugin/GlobalPluginService.go [hunk 8]
```diff
@@ -147,14 +159,35 @@ func (impl *GlobalPluginServiceImpl) GetAllGlobalVariables() ([]*GlobalVariable,
...
+		switch stageTypeReq {
+		case repository.CI_STAGE_TYPE:
+			stageType = repository.CI
+		case repository.CD_STAGE_TYPE:
+			stageType = repository.CD
+		case repository.CI_CD_STAGE_TYPE:
+			stageType = repository.CI_CD
+		default:
+			return nil, errors.New("stage type not recognised, please add valid stage type in query parameter")
+		}
```
Comment (kripanshdevtron) [inline pkg/plugin/GlobalPluginService.go:185]: create a util func of this

File: pkg/plugin/GlobalPluginService.go [hunk 9]
```diff
@@ -294,3 +301,1107 @@ func getVariableDto(pluginVariable *repository.PluginStepVariable) *PluginVariab
...
+			return nil, err
+		}
+		return pluginData, nil
+	case 2:
+		pluginData, err := impl.deletePlugin(pluginDto, userId)
+		if err != nil {
+			impl.logger.Errorw("error in deleting plugin", "err", err, "pluginDto", pluginDto)
+			return nil, err
+		}
+		return pluginData, nil
```
[the remark this answers is not in the document]
Comment (kripanshdevtron) [reply pkg/plugin/GlobalPluginService.go:354]: err handling and return statement can be common

File: pkg/plugin/GlobalPluginService.go [hunk 10]
```diff
@@ -294,3 +327,1353 @@ func getVariableDto(pluginVariable *repository.PluginStepVariable) *PluginVariab
...
+		Description: pluginReq.Description,
+		Type:        repository.PluginType(pluginReq.Type),
+		Icon:        pluginReq.Icon,
+		AuditLog: sql.AuditLog{
+			CreatedOn: time.Now(),
+			CreatedBy: userId,
+			UpdatedOn: time.Now(),
+			UpdatedBy: userId,
+		},
+	}
```
Comment (kripanshdevtron) [inline pkg/plugin/GlobalPluginService.go:407]: this could be shifted to bean folder

File: pkg/plugin/GlobalPluginService.go [hunk 11]
```diff
@@ -294,3 +327,1353 @@ func getVariableDto(pluginVariable *repository.PluginStepVariable) *PluginVariab
...
+	pluginStageMapping := &repository.PluginStageMapping{
+		PluginId:  pluginMetadata.Id,
+		StageType: pluginStage,
+		AuditLog: sql.AuditLog{
+			CreatedOn: time.Now(),
+			CreatedBy: userId,
+			UpdatedOn: time.Now(),
+			UpdatedBy: userId,
+		},
```
Comment (kripanshdevtron) [inline pkg/plugin/GlobalPluginService.go:428]: can create `NewDefaultAuditLog` func and can be used at many places

File: pkg/plugin/GlobalPluginService.go (hunk 10 above)
Comment (kripanshdevtron) [inline pkg/plugin/GlobalPluginService.go:407]: this could be shifted to bean folder

**Review by kripanshdevtron — APPROVED** [review]
